### ISSUE TITLE
Automatically add template variable as annotation tags

### ISSUE THREAD

**Issue by MattePalte** [grafana/grafana#24674]
**What would you like to be added**: These are the steps that I want to able to do:
- Select a specific panel in my dashboard. This panel has multiple time series of different objects (ca 80 - 200), all of the same type, a sort of swarm. I have a template variable to select (with checkbox) which object (via a name identifier) that I want to visualise.
- I select only one (or a subset) of those objects, because I want to add an annotation for them only. This annotation is representing an anomaly, it is an interval anomaly. 
- I want to be automatically add in the annotation's tags the identifier names of the objects. 
- Later I want to query all the anomalies on the specific object and see them in another dashboard.

**Why is this needed**: it is useful because it can help  in the very general task of annotating a time series dataset for the application of machine learning algorithms.

This is a very crucial task in all the major industrial setting in which data are collected through sensor on many machine ( monitoring of health of train track, power plant, servers).

In the use case I am working, we want to put in place a pipeline for annotating anomalies on hypervisors/servers, and this feature will help in making the annotation feature really usable and scalable in this scenarios. At the moment the only possible approach is that the annotator copy by hand the name identifier of the objects involved, that will require a lot more of (tedious) work that can be avoided by adding this simple feature.

In addition, this feature links very well to the other feature that lets user query based on annotation that are represented in a template variable (https://grafana.com/docs/grafana/latest/reference/annotations/#query-by-tag). It will be very easy to retrieve all the annotation on a specific machine with anomaly.

If you have alternative scalable approach please share it.
Thanks in advance

**Comment by MattePalte** [comment]
There is a growing interest in the annotation feature as witnessed by discussion such as this:
https://datascience.stackexchange.com/questions/38080/interactive-labeling-annotating-of-time-series-data

Grafana has really good characteristic in this regard and can compete also with specialised software. With the proposed feature Grafana will support a unique feature that will lead to a more solid annotation ecosystem with also the annotation part that can include tags.

Since at the moment we have template variable and tag that are very well integrated for QUERY STEP but not during the actual ANNOTATION STEP.

Thanks in advance

**Comment by lastlegion** [comment]
I agree this would be a great use case. Working on something very similar and running into the same issue.

**Comment by vkuznet** [comment]
+1

**Comment by MattePalte** [comment]
Thanks for the interest guys, we have been working on a patch for this functionality and it is impacting just the client-side JavaScript. We automatically add template variables as tags of the annotations and also add two buttons to trigger this behaviour. Therefore we decided to go for two customised buttons but we would ideally make this configurable by each organisation. 
<img width="881" alt="path_two_buttons" src="https://user-images.githubusercontent.com/15269127/86898456-04588d00-c109-11ea-8cf4-b508f4443676.png">
We just modified few lines of code, have a look here: https://gitlab.cern.ch/cloud-infrastructure/data-analytics/-/commit/794a7d79cc860cf5b1b545a5ae4cc255b6c53ea9 .

**Comment by vellala2000** [comment]
Adding variables as tags to annotations is a great idea. we also need to a way to filter annotations by dashboard variables when presenting.

**Comment by MattePalte** [comment]
@vellala2000 Thanks for the support! What you are suggesting is exactly the natural consequence, and actually it is already present as feature, it is present in the dashboard setting when you decide which annotation to display, maybe someone can jump in with a more precise description on how to find them if you can't find this functionality.

**Comment by vellala2000** [comment]
i am not able to find any such option to be able to control tags based on variable values at least for grafana data source. below is the screenshot. we are using Grafana v7.3.7.  However i see what you are referring, we can use query and variables to control the annotations when using external data sources.

![image](https://user-images.githubusercontent.com/3084986/110623222-90282a80-816a-11eb-9ec2-441ddf0ecb93.png)
![image](https://user-images.githubusercontent.com/3084986/110624017-a387c580-816b-11eb-9c26-cf97101f64fd.png)

**Comment by NerijusRazvodovskis** [comment]
Any updates on this one?

**Comment by MattePalte** [comment]
I am planning to work on it and I want to integrate it an appropriate way.

Our previous clearly temporary [patch](https://github.com/grafana/grafana/issues/24674#issuecomment-655385423) worked on the JavaScript code, now I want to implement the change directly in the TypeScript backbone, so it is stable.

For the **automatic addition of the template variables as annotation tags** I located this area of the project to work on:
https://github.com/grafana/grafana/tree/main/public/app/features/annotations 
Q1. Do you confirm this is the right place or should I work differently?

Whereas, if I want to add the two buttons (anomaly or normal), I was thinking of a plugin:
Q2: is plugin the way to go or should I still aim at a general core change which adds the option to add n-buttons or dropdown box to the usual annotation window?

I am not familiar with the repo yet, so I tag some of those involved in the last change in that area (@hugohaggmark , @domasx2 , @kylebrandt ) so to get their valuable feedback and guide me. Thanks in advance

**Comment by hugohaggmark** [comment]
@MattePalte thank you for creating the issue before we move on I think we need to think about how this feature should work in the general case, so I'm going to involve @pkolyvas for this.

> Q1. Do you confirm this is the right place or should I work differently?

This is the not entirely true, some of the code is here but we have a new Graph plugin called Time Series Panel that uses the code here: https://github.com/grafana/grafana/blob/main/public/app/plugins/panel/timeseries/plugins/AnnotationEditorPlugin.tsx

The old Graph panel has been deprecated.

> Q2: is plugin the way to go or should I still aim at a general core change which adds the option to add n-buttons or dropdown box to the usual annotation window?

My preference would be to design the UI/UX so the user can choose to add template variable values to tags automatically without the added plugin complexity.

**Comment by MattePalte** [comment]
@hugohaggmark thanks for the prompt answer and feedback. 

Meanwhile we wait for the comment of @pkolyvas I post here the UI/UX idea I had.

Goal of the change:
Allow the user select which templated variable should be added as annotation tag in automatic when hitting the save button in the annotation tooltip.

![UI_1_annotation_dashboard_setting](https://user-images.githubusercontent.com/15269127/131577583-e3cf8a8c-948c-493f-a404-fc0ee548850b.png)

Then every time the user adds a new annotation in by clicking add in the tooltip here below, he or she will get a new annotation with description the one chose and as tag it will have automatically added two tags (following the example): 
- "template variable 2: current_value_of_template_variable_2"
- "template variable 5: current_value_of_template_variable_5"

![annotation_tooltip](https://user-images.githubusercontent.com/15269127/131577694-aba090dd-241a-4177-af67-fcdf41e517fc.png)

Waiting for your feedback on this proposal. If a good idea please redirect me to the appropriate code location, so I will start studying and experimenting on it.

Thanks again

**Comment by dktcl** [comment]
Whilst I'm not skilled enough to support development directly, I would be delighted to help with beta testing - it's a feature I would love!

**Comment by xadorfr** [comment]
Very interested in that feature !

**Comment by AlexSzczepaniak** [comment]
+1 

Also interested in this feature request. The initial Post by Matteo is the exact functionality I am looking for as well.

**Comment by helinmik** [comment]
+1 We would also be very interested in this type of feature

**Comment by mimre25** [comment]
Also very interested in this feature!

@MattePalte @hugohaggmark @pkolyvas any update on this?

**Comment by bbili** [comment]
Any update on this ?

**Comment by AndreyStarostin** [comment]
+1 Very interested in that feature! Is there any news on this feature?

**Comment by JohnBolander** [comment]
Properly annotating dashboards where variables allow you to select from different datasources is basically broken without without this feature - solely relying on the user to tag the (correct) data source. When will this be implemented?

**Comment by Kariton** [comment]
hey folks,

just want to make my "+1".

in the end i only want to "filter by" variable tags.
exactly like here: https://community.grafana.com/t/using-template-variable-to-populate-annotation-tag/53091

to reviewing the impact of ansible (re) configurations.
ansible has a so called `callback plugin` which adds annotations for executed runs.
those annotations are tagged like:  `['ansible','ansible_event_start', 'playbook.yml', '_hostname_']` (4 different tags)
where the `_hostname_` would be my desired variable tag e.g.: `${instance}` in the context of `node exporter`

adding `${instance}` as tag to the annotation settings threads this as fixed string.

the appropriate grafana variable `${instance}` is present and does work to filter within promQL.

will it be possible to use "multi value" variables e.g.: `hosta|hostb`?
might also be an use case to consider.

**Comment by algo-se** [comment]
+1

**Comment by nliebelt** [comment]
+1

**Comment by CptOrange16** [comment]
+1

**Comment by SMLoureiro** [comment]
+1

**Comment by nastynaz** [comment]
+1

**Comment by ChrisSchmidtMT** [comment]
+1

**Comment by yongzhang** [comment]
I tried a custom variable with values like `typeA,typeB` and filter by `type:$type` does not work today

**Comment by bobemoe** [comment]
I think this is my exact issue, although I’m not 100% sure.

My use case is a dashboard with a variable such as `$node`. I add a manual annotation while viewing `node-a`, but when I switch the variable to `node-b`, the annotation remains because it appears to be attached to the panel rather than the selected node/data.

This also seems related to older issues such as #9735, which appears to have been only partly implemented: annotations can be filtered by variables/tags, but the current variable value is not automatically stored when creating the annotation.
